File: Stripe/STPCoreViewController.m
//
//  STPCoreViewController.m
//  Stripe
//
//  Created by Brian Dorfman on 1/6/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

#import "STPCoreViewController.h"
#import "STPCoreViewController+Private.h"

#import "STPColorUtils.h"
#import "STPLocalizationUtils.h"
#import "STPTheme.h"
#import "UIBarButtonItem+Stripe.h"
#import "UINavigationBar+Stripe_Theme.h"
#import "UIViewController+Stripe_NavigationItemProxy.h"
#import "UIViewController+Stripe_ParentViewController.h"

// Note:
// The private class extension for this class is in
// STPCoreViewController+Private.h

@implementation STPCoreViewController

- (instancetype)init {
    return [self initWithTheme:[STPTheme defaultTheme]];
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        [self commonInitWithTheme:[STPTheme defaultTheme]];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        [self commonInitWithTheme:[STPTheme defaultTheme]];
    }
    return self;
}

- (instancetype)initWithTheme:(STPTheme *)theme {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        [self commonInitWithTheme:theme];
    }
    return self;
}

- (void)commonInitWithTheme:(STPTheme *)theme {
    _theme = theme;

    if (![self useSystemBackButton]) {
        self.cancelItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
                                                                        target:self
                                                                        action:@selector(handleCancelTapped:)];

        self.stp_navigationItemProxy.leftBarButtonItem = self.cancelItem;
    }
}

- (void)setTheme:(STPTheme *)theme {
    _theme = theme;
    [self updateAppearance];
}

- (void)createAndSetupViews {
    // do nothing
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
    [self createAndSetupViews];
    [self updateAppearance];
}

- (void)updateAppearance {
    STPTheme *navBarTheme = self.navigationController.navigationBar.stp_theme ?: self.theme;
    [self.navigationItem.leftBarButtonItem stp_setTheme:navBarTheme];
    [self.navigationItem.rightBarButtonItem stp_setTheme:navBarTheme];
    [self.cancelItem stp_setTheme:navBarTheme];

    self.view.backgroundColor = self.theme.primaryBackgroundColor;

    [self setNeedsStatusBarAppearanceUpdate];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.view endEditing:YES];
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    STPTheme *navBarTheme = self.navigationController.navigationBar.stp_theme ?: self.theme;
    return ([STPColorUtils colorIsBright:navBarTheme.secondaryBackgroundColor]
            ? UIStatusBarStyleDefault
            : UIStatusBarStyleLightContent);
}

- (void)handleCancelTapped:(__unused id)sender {
    if ([self stp_isAtRootOfNavigationController]) {
        // if we're the root of the navigation controller, we've been presented modally.
        [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];

    } else {
        // otherwise, we've been pushed onto the stack.
        [self.navigationController popViewControllerAnimated:YES];
    }
}

- (BOOL)useSystemBackButton {
    return NO;
}

@end

File: Stripe/STPCoreViewController.swift
//
//  STPCoreViewController.swift
//  Stripe
//
//  Created by Brian Dorfman on 1/6/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import UIKit

/// This is the base class for all Stripe view controllers. It is intended for use
/// only by Stripe classes, you should not subclass it yourself in your app.
/// It theming, back/cancel button management, and other shared logic for
/// Stripe view controllers.
public class STPCoreViewController: UIViewController {
    /// A convenience initializer; equivalent to calling `init(theme: STPTheme.defaultTheme)`.
    @objc
    public convenience init() {
        self.init(theme: STPTheme.defaultTheme)
    }

    /// Initializes a new view controller with the specified theme
    /// - Parameter theme: The theme to use to inform the view controller's visual appearance. - seealso: STPTheme
    @objc public required init(theme: STPTheme?) {
        super.init(nibName: nil, bundle: nil)
        commonInit(with: theme)
    }

    /// Passes through to the default UIViewController behavior for this initializer,
    /// and then also sets the default theme as in `init`
    @objc public required override init(
        nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?
    ) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        commonInit(with: STPTheme.defaultTheme)
    }

    /// Passes through to the default UIViewController behavior for this initializer,
    /// and then also sets the default theme as in `init`
    @objc public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        commonInit(with: STPTheme.defaultTheme)
    }

    private var _theme: STPTheme = STPTheme.defaultTheme
    @objc var theme: STPTheme {
        get {
            _theme
        }
        set(theme) {
            _theme = theme
            updateAppearance()
        }
    }
    @objc var cancelItem: UIBarButtonItem?

    /// All designated initializers funnel through this method to do their setup
    /// - Parameter theme: Initial theme for this view controller
    func commonInit(with theme: STPTheme?) {
        if let theme = theme {
            _theme = theme
        } else {
            _theme = .defaultTheme
        }

        if !useSystemBackButton() {
            cancelItem = UIBarButtonItem(
                barButtonSystemItem: .cancel,
                target: self,
                action: #selector(STPAddCardViewController.handleCancelTapped(_:)))
            cancelItem?.accessibilityIdentifier = "CoreViewControllerCancelIdentifier"

            stp_navigationItemProxy?.leftBarButtonItem = cancelItem
        }

        NotificationCenter.default.addObserver(
            self, selector: #selector(STPAddCardViewController.updateAppearance),
            name: UIContentSizeCategory.didChangeNotification, object: nil)
    }

    /// Called in viewDidLoad after doing base implementation, before
    /// calling updateAppearance
    func createAndSetupViews() {
        // do nothing
    }

    // These viewDidX() methods have significant code done
    // in the base class and super must be called if they are overidden
    /// :nodoc:
    @objc
    public override func viewDidLoad() {
        super.viewDidLoad()

        createAndSetupViews()
        updateAppearance()
    }
    /// :nodoc:
    @objc
    public override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        updateAppearance()
    }
    /// :nodoc:
    @objc
    public override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        view.endEditing(true)
    }

    /// Update views based on current STPTheme
    @objc func updateAppearance() {
        let navBarTheme = navigationController?.navigationBar.stp_theme ?? theme
        navigationItem.leftBarButtonItem?.stp_setTheme(navBarTheme)
        navigationItem.rightBarButtonItem?.stp_setTheme(navBarTheme)
        cancelItem?.stp_setTheme(navBarTheme)

        view.backgroundColor = theme.primaryBackgroundColor

        setNeedsStatusBarAppearanceUpdate()
    }

    /// :nodoc:
    @objc public override var preferredStatusBarStyle: UIStatusBarStyle {
        let navBarTheme = navigationController?.navigationBar.stp_theme ?? theme
        return STPColorUtils.colorIsBright(navBarTheme.secondaryBackgroundColor)
            ? .default
            : .lightContent
    }

    /// Called by the automatically-managed back/cancel button
    /// By default pops the top item off the navigation stack, or if we are the
    /// root of the navigation controller, dimisses presentation
    /// - Parameter sender: Sender of the target action, if applicable.
    @objc func handleCancelTapped(_ sender: Any?) {
        if stp_isAtRootOfNavigationController() {
            // if we're the root of the navigation controller, we've been presented modally.
            presentingViewController?.dismiss(animated: true)
        } else {
            // otherwise, we've been pushed onto the stack.
            navigationController?.popViewController(animated: true)
        }
    }

    /// If you override this and return YES, then your CoreVC implementation will not
    /// create and set up a cancel and instead just use the default
    /// UIViewController back button behavior.
    /// You won't receive calls to `handleCancelTapped` if this is YES.
    /// Defaults to NO.
    func useSystemBackButton() -> Bool {
        return false
    }
}
